### ISSUE TITLE
Fix corrupt terminal output for list of warnings and errors

### ISSUE THREAD

**Issue by Huji** [stylelint/stylelint#3200]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

If you have a piece of CSS like this:

```
table.some_class {
  border-style: none;
}
```

It will fail the stylelint test. It would pass if you change it to:

```
table.some_class {
  border-style: none none none none;
}
```
This makes me think that stylelint is enforcing all borders to be specified, which is not what most (any?) browsers require.

This also passes the test:

```
table.some_class {
  border: 0;
}
```

But this fails:

```
table.some_class {
  border: none;
}
```

> Which rule, if any, is this issue related to?

Unfortunately the output of `npm test` doesn't specify which rule is triggered.

> What stylelint configuration is needed to reproduce this issue?

Unsure.

> Which version of stylelint are you using?

8.2.0

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Through `npm test` which in turn runs `grunt test`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

Test fails, see https://integration.wikimedia.org/ci/job/mwgate-npm-node-6-docker/25835/console for instance (the code is in https://gerrit.wikimedia.org/r/#/c/411593/19)

**Comment by ntwb** [comment]
Both code examples you supplied are identical, was that intentional?

Also reading your logs provides no reason for the stylelint failure, rather than posting the results from your CI what is the error shown when you run stylelint locally on your computer?

```
17:35:13 Running "stylelint:all" (stylelint) task
17:35:14 
17:35:14 modules/ext.abuseFilter.css
17:35:14  .. .  .  ...                                             ...                                 
17:35:14   . .  .                                                                                      
17:35:14     .  .
17:35:14 
17:35:14 ✖ 1 problem
17:35:14 
17:35:14 Warning: Task "stylelint:all" failed. Use --force to continue.
```

**Comment by Huji** [comment]
I ran stylelint on my computer through `npm test` and again, got no error returned, just a failure. I've updated the task above, and am trying to figure out how to run stylelint locally to get the error to actually show.

**Comment by jeddy3** [comment]
> I've updated the task above, and am trying to figure out how to run stylelint locally to get the error to actually show.

It sounds like something else is amiss here. Can you create a minimal reproducible test repo, please?

Just a `package.json`, `.stylelintrc` and single `test.css` should suffice.

The fact there is one error but it's not showing in the terminal makes me think it could be an issue with the particular terminal/OS and perhaps some special characters we use within the [default string formatter](https://github.com/stylelint/stylelint/blob/master/lib/formatters/stringFormatter.js). 

Can you try using the JSON formatter as that'll help narrow down where the problem might be e.g. [using grunt](https://github.com/wikimedia/grunt-stylelint#example-full-config):

```
// Project configuration.
grunt.initConfig( {
  stylelint: {
    options: {
      formatter: 'json','
    },
    src: [
            'src/**/*.{css,less,scss}'
        ]
    }
}
```

(FYI, there are no rules in stylelint that enforce writing out all the values of a shorthand property).

**Comment by Huji** [comment]
Well, I never use grunt except for the projects on which it is already configured so it will need a bit of learning on my end to be able to do so, but let's give it a try!

I created https://github.com/Huji/stylelinttest and tried to make it look close to the CI environment in which the error occurred. When I run `grunt` here, I get this error though:

```
Warning: Running stylelint failed
Error: No configuration provided for /home/huji/Desktop/csstest/test.css
  at module.exports (/home/huji/Desktop/csstest/node_modules/stylelint/lib/utils/configurationError.js:8:28)
  at /home/huji/Desktop/csstest/node_modules/stylelint/lib/getConfigForFile.js:46:15
 Use --force to continue.
```

what am I missing?

**Comment by jeddy3** [comment]
@Huji Thanks for following up with the test repo.

> Error: No configuration provided for /home/huji/Desktop/csstest/test.css

And [from the docs](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#configuration): 

> The linter _expects a configuration object_. You can either craft your own config or extend an existing one.

So, you'll need to add a `.stylelintrc` file, with the following content, to your test repo:

```
{ "extends": "stylelint-config-wikimedia" }
```

**Comment by Huji** [comment]
I added the .stylelintrc file as yo usuggested, and now this is the error that I get when I run grunt:

```
Running "stylelint:all" (stylelint) task

test.css
 . .  .  ...                                             ...                                 
 . .  .                                                                                      
 . .  .                                                                                      
 . .  .  ...                                             ...                                 
 . .  .                                                                                      
 . .  .

✖ 2 problems

Warning: Task "stylelint:all" failed. Use --force to continue.

Aborted due to warnings.
```

**Comment by jeddy3** [comment]
Very unusual. Is that the output from your CI or from your local machine? What is the OS of the machine?

Let's try to narrow down where the failure might be happening, by ruling out the Grunt plugin.

Can you please run stylelint via the CLI, rather than the Grunt plugin:

```
./node_modules/.bin/stylelint "*.css"
```

Do you get the same corrupted output?

**Comment by Huji** [comment]
@jeddy3 that was the output of CI. Running stylelint through CLI generated useful output:

```
 2:5  ✖  Expected indentation of 1 tab                   indentation                         
 2:5  ✖  Unexpected value "none" for property "border"   declaration-property-value-blacklist
```

Our focus is on the latter. It claims "none" is not an expected value for "border". I looked up W3C CSS2 documentation on https://www.w3.org/TR/2011/REC-CSS2-20110607/box.html#border-properties and learned that `border` is supposed process what is given to it in the following order: `<border-width> <border-style> <'border-top-color'>` and "none" is in fact not allowed for border-width, so indeed stylelint is correct in refusing to accept it, and browsers are just being too generous.

Closing this issue as invalid. Was very educational for me.

**Comment by hudochenkov** [comment]
@Huji `border: none` is a valid CSS. Because `border-style` is `none` by default.

**Comment by Huji** [comment]
Is "border: solid" valid CSS? Can you just simply skip some parts of a shorthand notation? I thought you cannot.

**Comment by hudochenkov** [comment]
@Huji yes, you can https://codepen.io/hudochenkov/pen/YaqKrm?editors=1100

For shorthand it doesn't matter if all values are present and order of these values.

```css
border: 1px solid black;
border: solid black 1px;
```

They are the same for a browser.

**Comment by Huji** [comment]
@hudochenkov whether browsers render it is one thing, whether W3C standard allows it is another. My understanding of reading W3C standard linked above was that order matters, and that `border-width` does not necessarily have a default value. I will leave this open for others' interpretation.

**Comment by hudochenkov** [comment]
It's also true for specification. And you can test in https://jigsaw.w3.org/css-validator/

> Value: | [ \<border-width> \|\| \<border-style> \|\| \<'border-top-color'> ]

`[ ]` means that what inside is optional. `||` means “or”. All together it means, that they all optional and could be in any order.

**Comment by jeddy3** [comment]
I'm going to close the issue as I believe the original corrupt terminal problem was proven (because the CLI works fine) not be related to stylelint itself.

I believe there are two things to clarify though:

1. It seems the corrupt terminal issue is likely connected to [grunt-stylelint](https://github.com/wikimedia/grunt-stylelint) and, @Huji, I'm sure they'd appreciate you raising an issue in their issue tracker so the problem is on their radar.

> so indeed stylelint is correct in refusing to accept it

2. stylelint itself isn't concerned with the correct (or best) value of the `border` property. The rule producing the violation is [declaration-property-value-blacklist](https://github.com/stylelint/stylelint/blob/master/lib/rules/declaration-property-value-blacklist/README.md). This is a rule that lets _configuration authors specify which values are disallowed for particular properties_. In this instance, it is the authors of [stylelint-config-wikimedia](https://github.com/wikimedia/stylelint-config-wikimedia/blob/master/index.js#L63) who have decided to disallow, in their code base, the value `none` for the `border` property. Providing rules to give _configurations authors the power to enforce what they believe to be the best values_ was a deliberate design decision. So, long story short, any discussion and/or issue with this configuration should probably happen in the [configuration author's issue tracker](https://github.com/wikimedia/stylelint-config-wikimedia/issues). 

Feel free to reopen the issue if you believe there's something more to do in stylelint itself.

**Comment by Volker-E** [comment]
The dotted output seems to have it's origin in grunt-stylelint as @jeddy3 already pointed out: https://github.com/wikimedia/grunt-stylelint/issues/49
